secret-sharing/src/vss/matrix.rs

```diff
@@ -0,0 +1,184 @@
+use group::{ff::PrimeField, Group, GroupEncoding};
+
+use super::polynomial::BivariatePolynomial;
+
+/// Verification matrix over a cryptographic group.
+///
+/// The verification matrix is computed as the element-wise scalar product
+/// of a matrix `b` representing the coefficients of a bivariate polynomial
+/// and a group generator `G`.
+///
+/// ```text
+/// M = [b_{i,j} * G]
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VerificationMatrix<G>
+where
+    G: Group + GroupEncoding,
+{
+    /// The number of rows in the verification matrix, determined by
+    /// the degree of the bivariate polynomial in the x variable from
+    /// which the matrix was constructed.
+    pub rows: usize,
+    /// The number of columns in the verification matrix, determined by
+    /// the degree of the bivariate polynomial in the y variable from
+    /// which the matrix was constructed.
+    pub cols: usize,
+    /// The verification matrix elements, where `m[i][j]` represents
+    /// the element `b_{i,j} * G`.
+    pub m: Vec<Vec<G>>,
+}
+
+impl<G> VerificationMatrix<G>
+where
+    G: Group + GroupEncoding,
+{
+    /// Constructs a new verification matrix from a given bivariate polynomial.
+    pub fn new<Fp>(bp: &BivariatePolynomial<Fp>) -> Self
+    where
+        Fp: PrimeField,
+        G: Group<Scalar = Fp> + GroupEncoding,
+    {
+        let rows = bp.deg_x + 1;
+        let cols = bp.deg_y + 1;
+        let mut m = Vec::new();
+        for ai in bp.b.iter() {
+            let mut mi = Vec::new();
+            for aij in ai.iter() {
+                mi.push(G::generator() * aij)
+            }
+            m.push(mi)
+        }
+
+        Self { rows, cols, m }
+    }
+
+    /// Returns true if and only if `m_{0,0}` is the identity element
+    /// of the group.
+    pub fn is_zero_hole(&self) -> bool {
+        self.m[0][0] == G::identity()
+    }
+
+    /// Returns the byte representation of the verification matrix.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let cap = Self::byte_size(self.rows, self.cols);
+        let mut bytes = Vec::with_capacity(cap);
+        let deg_x = (self.rows - 1) as u8;
+        let deg_y = (self.cols - 1) as u8;
+        bytes.extend([deg_x, deg_y].iter());
+        for mi in &self.m {
+            for mij in mi {
+                bytes.extend_from_slice(mij.to_bytes().as_ref());
+            }
+        }
+
+        bytes
+    }
+
+    /// Attempts to create a verification matrix from its byte representation.
+    pub fn from_bytes(bytes: Vec<u8>) -> Option<Self> {
```

Comment:
Would be good to have some fuzzing for the (de)serialization functions. For an example see the runtime package which uses hongfuzz.

Reply:
Added. Not sure how sdk runs `cargo hfuzz run ...` :shrug:

Reply:
One needs to run it manually currently.